feat: Replay IO traces on estimator for statistics

jakmeier opened the pull request:
This introduces the ability to replay IO traces on the estimator.
Example usage in `runtime/runtime-params-estimator/README.md`.

While it is not used for estimations, yet, it is already a subcommand of
the estimator. The idea is that for #7440 a lot of code can be reused.
Further, replaying on the estimator is required for #7058 and #7059.

jakmeier commented:
This is mostly code I used already for 1-2 months for performance debugging for a variety of different problems. Actually, it's only a selected subset of the code, pulled out to (finally) create a PR that introduces a IO trace replaying mechanism to the estimator. The code I didn't include is not ready, yet. And I realised I will probably not get around finishing that code any time soon. So it seems to be better to have the part that is ready reviewed and merged.

@matklad I picked you as a reviewer, hoping you might have some ideas on how to add tests. (So far there are none.) I was thinking about adding a test trace in a separate file and maybe use snapshot testing to check the output for each of the available commands. But I am not really sure.

Also, this is not the highest of priority to get it merged IMO. Feel free to put it aside until you have a moment to look though it, as it is a relatively large PR introducing a completely new concept to the estimator.

matklad commented:
Running these on hard-coded example seems like the right way to test it! You'll want to redirect output from `stdout` to a generic `dyn Write` to make this nice though.

jakmeier commented:
Updated this with tests and improved the output a bit. I think it's ready for a complete review now.

@matklad do you want to review this? Or should I ask someone more from the storage team?

matklad commented:
>Or should I ask someone more from the storage team?

Good call! I'd be more than happy to not learn more about this infra :D

jakmeier commented:
Hey @Longarithm, do you have some time to review this? It's not urgent at all, this should be a very low priority review. But it is a tool set that was really useful for me during Sweatcoin analysis and I think we should merge it in, perhaps add some documentation, and hopefully then it becomes useful for more people.

Quick overview:
- `neard view-state` has an option --record-io-trace since #7052
- IO traces are the input to this tool, which is part of the estimator (I have plans of using such traces for IO gas estimation)
- This PR includes real IO traces as test input, e.g. `runtime/runtime-params-estimator/res/75220100-75220101.s0.io_trace`
- The new tool reads the IO traces and summarizes statistics from it
- Several commands are available. Test output for each command is available.
    - `ReceiptDbStats`, per receipt DB access counts, `runtime/runtime-params-estimator/src/snapshots/runtime_params_estimator__replay__tests__ReceiptDbStats-75220100-75220101.s0.io_trace.snap`
    - Similar naming scheme for all other commands (`ChunkDbStats`, `CacheStats`, `ReceiptCacheStats`, `ChunkCacheStats`, `GasCharges`)
    - `GasCharges` can be interesting to point out, it checks DB accesses that we charge gas for vs DB accesses that are "free" because they are outside function calls

jakmeier commented:
Given our plans on how to benchmark flat storage, the priority of this has increased. We will need this as a base for replaying IO traces on RocksDB. (The code for replay on RocksDB is not included here but it builds on the same infrastructure.)

@Longarithm  Could you please review it some time this week? Or if you don't have time let me know and I will look for someone else to review it.

Longarithm commented on runtime/runtime-params-estimator/src/replay/fold_db_ops.rs:
```diff
@@ -0,0 +1,263 @@
+    /// Labels at which to break aggregation.
+    fold_anchors: Vec<String>,
+    /// Fields to display on each output, if they are available in the trace.
+    printed_fields: Vec<String>,
```
Can we store it as a mapping from anchor to printed fields?

I see that it doesn't impact our usecases, but I think it makes code more readable. Initially I got confused why do we have a duplicated printed field `fn receipts(self)`, and printed fields are connected to specific anchor anyway.

jakmeier replied:
yes that make sense, I didn't even notice the duplicate output, that was a bug too^^
anyway, using a map makes more sense

Longarithm commented on runtime/runtime-params-estimator/src/replay/fold_db_ops.rs:
```diff
@@ -0,0 +1,263 @@
+        self.states.push(new_state);
+    }
+
+    fn pop_state(&mut self) -> State {
```
Is it intended that outer states don’t aggregate stats from inner states? I’m a bit confused why it is the case, because I expected the top-level state to aggregate everything.

jakmeier replied:
Yes that is intended. My thinking is this: A per-receipt output is mostly interested in the output for each receipt. But there is also some stuff that happens in between, which I called the top-level. Knowing how much is not associated with a receipt seems more interesting in this output than knowing the total.
Also, the total can still be retrieved by calling another command.

Does that make sense? Or would you say it makes more sense to use aggregate by default?

Longarithm replied:
It makes sense, thank you!

jakmeier commented on runtime/runtime-params-estimator/src/replay/fold_db_ops.rs:
```diff
@@ -0,0 +1,263 @@
+            cache_stats.eval_db_op(op, size);
+        }
+
+        self.check_indent(out, indent)?;
```
that's a bug, good catch!

jakmeier commented on runtime/runtime-params-estimator/README.md:
```diff
@@ -30,3 +30,55 @@ Use this tool to measure the running time of elementary runtime operations that
+do
+  target/release/neard \
+    --record-io-trace=75220100-75220101.s${shard}.io_trace view-state \
+    apply-range --start-index 75220100 --end-index 752201 \
```
fixed

jakmeier commented on runtime/runtime-params-estimator/src/replay/fold_db_ops.rs:
```diff
@@ -0,0 +1,263 @@
+use std::collections::BTreeMap;
+use std::io::Write;
+
+const EMPTY_STATE: &str = "states must never be empty";
```
fixed

jakmeier commented on runtime/runtime-params-estimator/src/replay/fold_db_ops.rs:
```diff
@@ -0,0 +1,263 @@
+}
+
+#[derive(Default)]
+struct State {
```
Comment added, let me know if you think it should be rephrased.

jakmeier commented on runtime/runtime-params-estimator/src/replay.rs:
```diff
@@ -0,0 +1,280 @@
+    }
+}
+
+trait Visitor {
```
Also here, comment added, let me know if you think it should be rephrased.

Longarithm commented on runtime/runtime-params-estimator/src/replay/fold_db_ops.rs:
```diff
@@ -0,0 +1,268 @@
+
+    fn skip(&mut self, indent: usize) -> bool {
+        if let Some(skip_indent) = self.skip_indent {
+            if skip_indent >= indent {
```
Hm, shouldn't it be `if skip_indent > indent`? Perhaps it doesn't matter for our usecases, but I imagine there can be many ops with the same indent to skip, and the current code skips only one and don't skip others.

I would cover the expected behaviour in some test, which could be either synthetical or include receipts with receiver which must be skipped.

Longarithm commented on runtime/runtime-params-estimator/src/replay/fold_db_ops.rs:
```diff
@@ -0,0 +1,268 @@
+        key: &[u8],
+        col: &str,
+    ) -> anyhow::Result<()> {
+        self.check_indent(out, indent)?;
```
I took another look and now I don't understand why here `check_indent` comes before `skip`, it is the same for `eval_storage_op`, but for `eval_label` the order is different.

Could you explain this logic? Maybe `fn skip` is a good place to comment this behaviour. Similarly, I think we need some test which fails for the case when order is incorrect.

jakmeier commented:
@Longarithm I changed how the account filter works substantially. Most importantly, data that belongs to no account will be excluded from the reports when an account filter is active.
For that, I inverted the skipping behavior. Instead of skipping every account that is not in the filter list, we now skip by default and enable statistics collection once we find a matching account. This makes much more sense to me.^^ Hopefully that makes the code clearer as well.

Also added tests for this.

Longarithm commented on runtime/runtime-params-estimator/src/replay/fold_db_ops.rs:
```diff
@@ -0,0 +1,284 @@
+    ///
+    /// Used to apply filters and skip some parts of the trace.
+    /// Default is 0, which collects everything.
+    min_indent: usize,
```
Can we name it to `max_indent` or `indent_limit` because this is an upper bound for indent, not a lower bound?

jakmeier replied:
Hm, no I think it is the minimum indentation for a line to be evaluated. If the minimum is 0, it means everything is included. If the minimum is usize::MAX, nothing is included.

It could also be seen as the upper bound on what is being skipped. But I find that less intuitive. I updated the comment to describe in terms of what is included instead of what is filtered.

Longarithm commented on runtime/runtime-params-estimator/src/replay/fold_db_ops.rs:
```diff
@@ -0,0 +1,284 @@
+    /// Check if indentation has gone back enough to pop current state or reset filter.
+    ///
+    /// Call this before `skip()` to ensure it uses the correct `min_indent`.
+    fn check_indent(&mut self, out: &mut dyn Write, indent: usize) -> anyhow::Result<()> {
```
I think the naming could be improved here. `check_indent` doesn't just check indentation, it modifies stack of states, and `skip` doesn't give intuition what is actually skipped.

What do you think about `check_state` (`check_current_state`?) and `skip_eval`?

jakmeier replied:
yes my names were really bad here :)
I did the following changes:

`check_indent` -> `update_state`
`skip` -> `skip_eval`

Longarithm reviewed (approved):
LGTM, leaving 2 more comments about naming.
